In [19]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_regression
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier


In [13]:
# Import the data

df = pd.read_csv(Path('Resources/lending_data.csv'))
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [15]:
# Split the data into X_train, X_test, y_train, y_test

X = df.drop(columns = ["loan_status"], axis=1)

y = df[["loan_status"]]

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


In [16]:
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [17]:
y.head()

loan_status
0            0
1            0
2            0
3            0
4            0

In [20]:
# Train a Logistic Regression model print the model score
classifier = LogisticRegression()
classifier.fit(X_train, y_train)
print(f"Training Data Score: {classifier.score(X_train, y_train.values.ravel())}")
print(f"Testing Data Score: {classifier.score(X_test, y_test.values.ravel())}")

Training Data Score: 0.9921240885954051
Testing Data Score: 0.9918489475856377


/opt/anaconda3/envs/PythonData38/lib/python3.8/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [21]:
# Train a Random Forest Classifier model and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

/var/folders/53/8v7yl9_11zqgtgh__mhvy_5w0000gn/T/ipykernel_39527/3806078898.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)


Training Score: 0.9975409272252029
Testing Score: 0.991642591828312


## Results and Predictions

The machine learning algorithms random forest and logistic regression are used for various disciplines for classification and regression purposes. 

Logistic regression is a parameter based model and random forest is a non-parametric model. Logistic regression will perform better since it mapsqualitative or quantitative imput features to a target variable. It is used to predict financial, biological or socialogical data in supervised machine learning. 

Random Forest however is ensemble-based learning algorithm, comprised of n collections of de-correlated decision trees.

Results: 

Logistic Regression model training data score = 0.9921240885954051
Logistic Regression model testing data score = 0.9918489475856377

Random Forest Classifier model training data score = 0.9975409272252029
Random Forest Classifier model testing data score = 0.991642591828312